### PR TITLE
Add @materialize macro

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
The materialize macro is a utility function from GLVisualize which seems well-suited for this package. It splats a dictionary to variables in a similar way to how your `@unpack` splats a type to variables.

### CODE REVIEW

File: src/Parameters.jl [hunk 1]
```diff
@@ -460,4 +460,25 @@ macro pack(arg)
...
+
+```
+@materialize a, b, c = dict
+```
+
+"""
+macro materialize(dict_splat)
+    keynames, dict = dict_splat.args
+    keynames = isa(keynames, Symbol) ? [keynames] : keynames.args
+    dict_instance = gensym()
```
Comment (mauro3) [inline src/Parameters.jl]: Why is that needed?  Why not work with `$dict`?

**Comment by mauro3** [comment]
Thanks!  I think at some stage the `@(un)pack` macros and this macro should go to a separate package, but until that time this fits well.

Would you mind adding tests and a bit to the docs?

File: src/Parameters.jl (hunk 1 above)
Comment (ChrisRackauckas) [inline src/Parameters.jl]: @SimonDanisch

**Comment by ChrisRackauckas** [comment]
Yes. I'll try to get to that soon.

**Comment by ChrisRackauckas** [comment]
Added docs, improved the docstring, and added tests.

**Comment by mauro3** [comment]
Thanks!

I just noted that the syntax is not aligned with the `@unpack` syntax: 
```julia
@unpack v: a, b, c 
# vs
@unpack a, b, c = v
```
Arguably, the materialize syntax is more consistent with other Julia syntax.  Is the `@unpack` syntax more readable?  Let me know what you think.  They should be consistent, for sure.  

Also, whilst `@materialize` is a very cool name, more consistent and readable would be the dull `@unpackdict`.  Probably there is a more clever, unifying syntax.

File: src/Parameters.jl (hunk 1 above)
Comment (SimonDanisch) [inline src/Parameters.jl]: Oh sorry! Didn't see this... Yeah, that's for the case when the dict gets computed (which you don't want to do for every key access)!
I've actually never used the macro like that so far, though :D

Comment (mauro3) [inline src/Parameters.jl]: Ah, I see, clever.  Tnx.

**Comment by ChrisRackauckas** [comment]
I think matching the Julia syntax is better, and would prefer `@unpack` look like a Julia assignment. There should probably be a pack to go along with it?

I think `@unpackdict` is fine, I just used `@materialize` because it was what the name was from GLVisualize when Simon showed me this.

**Comment by mauro3** [comment]
Yes, `@pack` would be:
```
@pack v = a, b, c
```
Looks better too.

Considering the breaking syntax change, I think the best way forward would be to do the split of the Parameters package now into, say, `KWTypes.jl` and `Unpack.jl`.

**Comment by SimonDanisch** [comment]
`unpack` and `pack` seems like a nice pair ;) I think mirroring Julia syntax is the way to go!
I wouldn't go for appending `dict` to pack.
Pack would do this? : 
```Julia
@pack v = a, b, c
v == Dict(:a=>a, :b=>b, :c=>c)
```
Note that I also have `materialize!` which deletes the key. I find that also useful from time to time.
Also there is `gen_default`, which is currently quite specific to `GLVisualize`, but the generic form might also be interesting:
```Julia
@gen_default dict begin
    a = ...
    b = ...
end
-->
if haskey(dict, :a)
    dict[:a] = a
end
if haskey(dict, :b)
    dict[:b] = b
end
```
I incorporated a little DSL for conversions and attaching doc strings to the keys, which is probably more specific to plotting:
```Julia
@gen_default dict begin
    a = ... => Vector{Float32}, "This is kw_arg a and will get converted to Vector{Float32}"
    b = ...
end
```

**Comment by mauro3** [comment]
Adding the `dict` is because `@unpack` is for unpacking a type:
```
type T
  a
  b
end
t = T(1,2)
@unpack a, b = t
# expands to
a = t.a
b = t.b
```
So, I think two different macros are needed.  Or do you have an idea on how to unify those?

The `@gen_default` seems similar to the keyword types.  Maybe that could be unified too?

**Comment by SimonDanisch** [comment]
you could lower `@unpack` to something like:
```Julia
a = _unpack(t, :a)
b = _unpack(t, :b)
#with
_unpack(x, field) = getfield(x, field)
_unpack(x::Dict, key) = x[key]
```

**Comment by mauro3** [comment]
Now implemented here: https://github.com/mauro3/Unpack.jl . Comments welcome!

I still need to decide whether to split the package or not.

**Comment by ChrisRackauckas** [comment]
This can get closed then. Whether or not it goes into this package, it'll be the new implementation.

**Comment by mauro3** [comment]
This is now itagged in v0.5.0

**Comment by mauro3** [comment]
Turns out to get type-stability some `Val` tricks were needed: https://github.com/mauro3/Parameters.jl/blob/87ffb34adb3555fded8cfc2227d4defd22a38be7/src/Parameters.jl#L424
Although only works on Julia 0.5.
